[Common] add utilities for event selection based on data quality

aferrero2707 opened the pull request:
- definition of bits in the CCDB objects
- utility class to check a collision/bc based on a given list of bits

aferrero2707 commented:
@ddobrigk @ekryshen here is a draft proposal for the interface to the time-dependent quality flags in the CCDB.
I will put more comments and documentation once we agree on the implementation.

aferrero2707 commented:
@ddobrigk @ekryshen I have completed the work on the helper code for the quality-based event selection.

The code provides the definition of the bits associated to the quality flags, and an helper class to check a given collision or BC against a list of quality flags that are required not to be Bad.

The usage of the helper class (`QualityFlagsChecker`) is pretty simple.
Here are some examples of possible initializations of the quality checker class:
```
  // initialization from runlist lablel
  std::array<QualityFlagsChecker, 6> qualityFlagsCheckers{
    "CBT", "CBT_hadronPID", "CBT_electronPID", "CBT_calo", "CBT_muon", "CBT_muon_glo"
  };

  // initialization from runlist label, setting also second parameter (note the double outer brace)
  std::array<QualityFlagsChecker, 6> qualityFlagsCheckersWithZDC{{
    {"CBT", true}, {"CBT_hadronPID", true}, {"CBT_electronPID", true}, {"CBT_calo", true}, {"CBT_muon", true}, {"CBT_muon_glo", true}
  }};

  // default constructor, the checker must be initialized later by calling the initialize() method with proper parameters
  QualityFlagsChecker qualityFlagsCheckerMuonNoLimAcc;

  // initialization with a list of quality flags to be checked
  QualityFlagsChecker qualityFlagsCheckerCustom{ kFT0Bad, kITSBad, kMFTBad, kMFTLimAccMCRepr };
```

The check of the quality of a given collision is also very simple:
```
bool ok = qualityFlagsCheckerCustom(collision);
```

Let me know if the proposed interface is ok for you.

aferrero2707 commented:
@ddobrigk @ekryshen @f3sch I think the code is now ready for the final review, I have implemented all suggestions received in the past weeks.

f3sch reviewed (commented):
Hi @aferrero2707, nice work, minor comments things (prob. even overengineering the problem so feel free to ignore).
Also I noticed that you use uint64_t while the bitmap is uint32_t which you also showed in the presentation.

aferrero2707 commented on Common/CCDB/RCTSelectionFlags.h:
```diff
@@ -0,0 +1,220 @@
+class RCTFlagsChecker : public o2::utils::EnumFlags<RCTSelectionFlags>
+{
+ public:
+  RCTFlagsChecker() = default;
```
It is needed to allow the use of the default constructor, with a deferred initialization in the `init()` function of the analysis task.

aferrero2707 commented on Common/CCDB/RCTSelectionFlags.h:
```diff
@@ -0,0 +1,220 @@
+                  flags.end(),
+                  [this](const RCTSelectionFlags f) noexcept { set(f); }
+    );
+  }
```
Indeed, more compact.

aferrero2707 committed:
[Common] add utilities for event selection based on RCT flags

- definition of bits in the RCT CCDB objects
- utility class to check a collision/bc based on a given set of bits

aferrero2707 commented:
@f3sch I added your review comments, except for the default constructor which is actually needed if one wants to initialize an object without parameters, and use the `init()` function later on.

Regarding the cast to `uint64_t`, it was intended to make the code future-proof, and to make sure we do not end up mixing 32 and 64 bit words in the comparison. But maybe it's not needed?

f3sch commented:
@aferrero2707, regarding the uint would a simple `auto` suffice then it will change transparently when the bits are increased.

aferrero2707 commented:
> @aferrero2707, regarding the uint would a simple `auto` suffice then it will change transparently when the bits are increased.

@f3sch if it is guaranteed that the type returned by `rct_raw()` and the underlying type of the `RCTSelectionFlags` enum are always compatible, then I agree that `auto` is enough. Otherwise, I would keep the cast to `unit64_t` which is likely the wider bit representation we will use for those flags...

aferrero2707 commented:
@f3sch @jgrosseo @iarsene @ekryshen @ddobrigk @ktf I would propose to have a final review of this one, and merge if there are no more suggestions, as it would be very useful to get feedback for experienced analyzers.
Thanks!

ekryshen reviewed (approved):
@aferrero2707 I am not an expert in EnumFlags helpers but it looks good from the first glance.

ddobrigk reviewed (approved):
thanks a lot! Will merge for first tests - we can do fine tuning in a later PR.
